### ISSUE TITLE
Defining rules for updating existing images.

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#481]
Further to the discussion in #471 I would like to consider how we might go about updating existing images, primarily to improve the quality of an image. I envisage running a one-time background process the first time a new version of Magarena is started (and import chosen) which would silently update images based on a clear set of rules. It should be able to handle such cases as where someone has manually replaced an image with their own preferred image so that it is not automatically overwritten (eg. Russian edition of card image).

**Comment by lodici** [comment]
My initial thoughts on the rules to use -
```
for each image file in 'images\cards' and 'images\tokens'...
    if (local file size is equal to remote file size)
	    skip
    else
        - download remote image file so image sizes (**not file size**) can be compared.
	    - if remote image size is bigger than local
            update local image.
```

Please note that when comparing remote and local file sizes you cannot assume that a bigger size means better quality. I have many 480 x 680 images from ``mtgimages.com``, a number of which have a smaller file size than the same 312 x 415 image on ``magiccards.info``.

**Comment by lodici** [comment]
```
for each image file in 'images\cards' and 'images\tokens'...
    if (local file size is equal to remote file size)
	    skip
    else
        - download remote image file so image sizes (**not file size**) can be compared.
	    - if remote image size is bigger than local
            update local image.
        - if remote image size is the same as local
           - if remote file size is bigger than local
                update local image
```

**Comment by lodici** [comment]
The one question mark I have with the above approach is that it will mean many instances of Magarena hammering ``magiccards.info`` when checking file sizes after each new release. I am not sure whether that might be an issue or not with ``magiccards.info`` in terms of load. I have found that for 100 images it takes about 10 seconds to get the remote files sizes so although I think it is faster than actually downloading the image files it still obviously requires some work on behalf of ``magiccards.info``.

An alternative approach could be to automatically generate the list of images to be updated just prior to each new release. This could take the form of adding an update file to ``resources`` or adding/updating the ``image_updated`` property in the appropriate scripts. However, that is less flexible than treating each user's image set on its own merits.

**Comment by melvinzhang** [comment]
I have the same concern of running this on each install as it will take quite a while to run to check all the images. Using image_updated means we can run it and it can even be a separate program instead of integrating into the game engine itself.

**Comment by lodici** [comment]
I like the idea of using ``image_updated`` as it means no changes required to the current mechanism for updating and downloading images. It also means that we can manually tweak things if required.

My initial thoughts on how this might work if the process is to be automated -

Before each release run separate program that...

1. creates a list of remote image file sizes for each playable card - ``<NEW LIST>``.
2. if ``\resources\<OLD LIST>`` exists then
    1. compare ``<NEW LIST>`` to ``<OLD LIST>`` by running image update rules against each entry.
    2. if rules indicate image should be updated then set ``image_updated`` property in associated script.
3. add ``<NEW LIST>`` (replacing ``<OLD_LIST>`` if present) to ``\resources``.
4. commit.

By using a reference list under source control it should mean anyone of us can run the program and get consistent results. I think!

**Comment by lodici** [comment]
Alternatively, 

Before each release run separate program that...

1. creates a list of remote image file sizes for each playable card - ``<REF_LIST>``. We do this once to avoid the hammering of ``magiccards.info`` as mentioned above.
2. add ``<REF_LIST>`` (replacing ``<OLD_LIST>`` to ``\resources``.
3. commit.

On Magarena startup,

1. run rules described above but use ``<REF_LIST>`` to determine initial set of images to be downloaded.

**Comment by ShawnieBoy** [comment]
If you require a test - Magic Origins (ORI) is now better quality than it's first release on magiccards

**Comment by lodici** [comment]
Thanks for the heads up - that will be useful.

**Comment by ShawnieBoy** [comment]
I've just had a quick look myself, and as I expected, the properties of the images are identical. Dimension, file-size etc. Would have to be assessed by the human eye - or to have a batch re-download of particular sets when the quality is better.

**Comment by lodici** [comment]
> or to have a batch re-download option for particular sets when the quality is better.

I was thinking this as well - in fact I would see this a good first step while we figure out a foolproof automated approach (if indeed it is even possible). The explorer seems a good place to add a function that will let you re-download images for either the current selected image or the set of cards as defined by the filter.

**Comment by lodici** [comment]
We could also possibly use an alert to notify users when better quality images are available - _hey, Magic Origins (ORI) images have been updated, open the ORI set in explorer and click on the update images button_

**Comment by ShawnieBoy** [comment]
Both sound like great ideas :+1:

**Comment by melvinzhang** [comment]
The alert seems kind of tedious, if we use image_update to signal an update, the missing images alert will show and they can immediately download the new images.

**Comment by ShawnieBoy** [comment]
Adding the 'Image update' to a whole set's worth of script files is also quite tedious. This way it will enable users to only update the images they want, especially with non-English or other custom images involved.

**Comment by melvinzhang** [comment]
I think for overwriting, the images should be put in a different folder, eg images/override. We may still introduce image_update to fix wrong images and that will still break custom images.

**Comment by melvinzhang** [comment]
adding image_update can be automated as the values are all going to be the same, just need to indicate which cards are involved.

**Comment by lodici** [comment]
> I think for overwriting, the images should be put in a different folder, eg images/override

Yep, that would be how I would do it. It is simple enough to check the ``override`` folder first when retrieving a image.

**Comment by lodici** [comment]
> adding image_update can be automated as the values are all going to be the same, just need to indicate which cards are involved.

Using image_updated is the way to go I think for an automated update but I am still to be completely convinced we can come up with a full proof method for doing this,

In the meantime, having a manual function that will let you update a selected image or images in the explorer or deck editor would be useful even once and if an automated solution is implemented.

**Comment by lodici** [comment]
With commit  f88e3af Magarena will now use an image found in the optional ``<images>\custom`` directory instead of the default ``cards`` or ``tokens`` directories. By using a separate directory this should prevent any standard updates from overwriting the custom image.

**Comment by ShawnieBoy** [comment]
>Using image_updated is the way to go I think for an automated update but I am still to be completely convinced we can come up with a full proof method for doing this,

I feel as though some split in the importance where the image_updated property would be needed. There are times when, for a player, we should indicate 'You NEED to download this image, the game will not be working correctly otherwise' - a wrong image, old card text etc. And 'These images are available, you can download them if you want to.' different quality/version.

**Comment by melvinzhang** [comment]
I feel that a quality update should join the former list and be automatically included rather than requiring manual selection. Version update is debatable, in the case of the MM2 update, many of the card rarity were also modified and no longer match the previous image, in that case I think we should also automatically update the image so that card rarity in the card script and image matches.

**Comment by ShawnieBoy** [comment]
Good point about the rarity changes, and thinking about it, quality updates would probably lean more towards the enforced updating as well.

I guess with cards that are solely a change in version would aid new installs, and I guess with @lodici 's  manual update idea, we could mention it in the forums or update prompt that they are available.

**Comment by ShawnieBoy** [comment]
Oh, another thing - as eventually all cards will have an image_updated property, would adding a blank property to the scriptbuilder outputs be worthwhile?

**Comment by melvinzhang** [comment]
> would adding a blank property to the scriptbuilder outputs be worthwhile?
No, I don't think so. Color is the only property that is legal to be blank, I don't think we should add another one.

**Comment by lodici** [comment]
See also issue #501 which updates ``image_updated`` to work with images for missing cards.

**Comment by lodici** [comment]
One thing to be aware of with ``image_updated`` is that is has a granularity of a day. This means that if ShawnieBoy updates images on Monday then when I download first thing on Tuesday I will get these new images. But if ShawnieBoy then does another image update later on Tuesday I will not pick these up because ``imageDownloaderRunDate`` in ``general.cfg`` will have already been set to Tuesday's date when I ran the download in the morning.

I don't think it is a biggy. It should only affect those of us using the latest dev builds or source code. The fix is to delete the ``imageDownloaderRunDate`` and ``missingImagesDownloadDate`` from ``general.cfg``. A more permanent solution would be to include the time with ``image_updated`` but not sure it is worth the trouble.

**Comment by ShawnieBoy** [comment]
There is a bit of a bigger issue with the using of dates - We have seen many people installing 1.67 recently. These will have a date of images downloaded later than the image changes in 1.68.

This isn't much of an issue if the image hasn't got a change of image url, but if they have, they will not be downloaded when later updated to version 1.68 due to the last downloaded date.

**Comment by lodici** [comment]
Very good point. 

Possible fix ``#1`` : Run a batch script or java app that checks every script file and for every image link that has been updated in last month, update ``image_updated`` to the next release date.

Possible fix ``#2`` : Include the version with download date.

**Comment by ShawnieBoy** [comment]
Options 3 sounds like the best bet, as just today Commander 2015 is now in HQ - Time for some more image updating...

**Comment by lodici** [comment]
The above commit should work for most cases where someone is regularly updating versions - it should catch any scripts with an ``image_updated`` value that falls within the last 60 days. Less so if someone upgrades from a much later version (eg.  1.62 to 1.67) where the 60 days would not be enough. If I can work out how to get the version that is being imported then can calculate how many days to look back.

**Comment by ShawnieBoy** [comment]
#471 #777 and #837 are related here, as I think the rules defining card images should be updated to allow for multiple language support and crops.
Defining which sets allow for most compatibility will expand the number of alternate language images and crops available to download. I've already found that online-only sets only appear in English and have no crops available, the same goes for 'special' sets like commander.
The priority for English oracle text on card images isn't so high any more as we have the most up to date oracle already on the card scripts, and rendered with cardbuilder providing there's a crop available.
